### PR TITLE
tests: support running benches via custom test runner

### LINKED ISSUE #12820: Benches under tests crate cannot be run
### What did happened?

Due to this line: https://github.com/tikv/tikv/blob/0d9b7b33b7928be06016075329fe0b2c4fba25f5/components/test_util/src/runner.rs#L66
All benches under `tests` are forcefully converted to unit tests.

### PR DESCRIPTION
**Pull request by tabokie**
Signed-off-by: tabokie <xy.tao@outlook.com>

### What is changed and how it works?

Issue Number: Close #12820

What's Changed:

Run bench with test hook.

~Also introduce a new env variable that can fold benches into unit tests on demand. This is helpful for test framework that doesn't support benches yet (such as nextest).~

```commit-message
```

### Check List 

Tests 

- Unit test
- Manual test (add detailed scripts or steps below)

```bash
# cargo test --package tests -- --test bench_txn_store_rocksdb_inc
test storage::test_storage::bench_txn_store_rocksdb_inc_x100 ... ok
test storage::test_storage::bench_txn_store_rocksdb_inc ... ok
# cargo test --package tests -- --bench bench_txn_store_rocksdb_inc
test storage::test_storage::bench_txn_store_rocksdb_inc      ... bench:   1,626,586 ns/iter (+/- 373,337)
test storage::test_storage::bench_txn_store_rocksdb_inc_x100 ... bench:  44,463,530 ns/iter (+/- 9,107,088)
```

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by zhangjinpeng1987 — APPROVED** [review]
LGTM

**Review by BusyJay — COMMENTED** [review]
What about doc tests?

File: components/test_util/src/runner.rs [hunk 1]
```diff
@@ -31,42 +31,50 @@ impl TestHook for Nope {
...
         self.hook.teardown();
-        debug!("case end"; "name" => &self.name);
     }
 }
 
 /// Connects std tests and custom test framework.
-pub fn run_test_with_hook(cases: &[&TestDescAndFn], hook: impl TestHook + Send + Clone + 'static) {
+pub fn run_test_with_hook(
+    cases: &[&TestDescAndFn],
+    hook: impl TestHook + Send + Sync + Copy + 'static,
```
[the remark this answers is not in the document]
Comment (tabokie) [reply components/test_util/src/runner.rs:53]: Dynamic bench requires a `Fn`, but TestHook mutates itself and cannot be moved inside.

**Comment by tabokie** [comment]
@BusyJay Doc tests are not run in our CI either. But if it's needed, `--persist-doctests` is probably good enough.

File: components/test_util/src/runner.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (BusyJay) [reply components/test_util/src/runner.rs:53]: I don't get it. If someone wants to run bench, it should use command `cargo bench` right? What's the point of changing runner?

[the remark this answers is not in the document]
Comment (tabokie) [reply components/test_util/src/runner.rs:53]: `cargo bench` goes through the same logic as well. Currently there's just no way to run benches in `integrations` folder under debug mode. To me, the best solution would be only supporting test hook for tests, and panic if there is a bench and a non-empty hook.

[the remark this answers is not in the document]
Comment (BusyJay) [reply components/test_util/src/runner.rs:53]: > this solution to https://github.com/tikv/tikv/issues/12769?

How are they related? Even the master implementation can run the bench as test already?

[the remark this answers is not in the document]
Comment (tabokie) [reply components/test_util/src/runner.rs:53]: Nextest ignores bench cases, it's the only blocking issue right now.

[the remark this answers is not in the document]
Comment (BusyJay) [reply components/test_util/src/runner.rs:53]: Can you explain more how this PR solves the issue that "nextest ignores bench cases" but master won't? What's the details solution?

[the remark this answers is not in the document]
Comment (tabokie) [reply components/test_util/src/runner.rs:53]: > Also introduce a new env variable that can fold benches into unit tests on demand. This is helpful for test framework that doesn't support benches yet (such as nextest).

[the remark this answers is not in the document]
Comment (BusyJay) [reply components/test_util/src/runner.rs:53]: No, it explains nothing. Master's implementation already folds benches into unit tests no matter there is environment variables or not.

[the remark this answers is not in the document]
Comment (tabokie) [reply components/test_util/src/runner.rs:53]: It does not (except benches in tests crate). There is a difference between "folding benches into tests" and "treating benches as tests inside cargo test implementation".

[the remark this answers is not in the document]
Comment (tabokie) [reply components/test_util/src/runner.rs:53]: If you are referring to the raw test binary's handling of cases, then I agree benches are equivalent to tests. But what concerns us is how test framework thinks.

[the remark this answers is not in the document]
Comment (BusyJay) [reply components/test_util/src/runner.rs:53]: > except benches in tests crate

So your solution is to make the test runner as the default runner for all crates?

And the reason you introduce the change is to make sure after making it the default, all bench cases can still be bench?

[the remark this answers is not in the document]
Comment (tabokie) [reply components/test_util/src/runner.rs:53]: Yes. I do consider this approach less satisfying (especially the part where every new packages have to use this runner). I just filed a PR to see if nextest can treat benches the same way as cargo test (https://github.com/nextest-rs/nextest/pull/283).

[the remark this answers is not in the document]
Comment (BusyJay) [reply components/test_util/src/runner.rs:53]: The change as a fix for unable running benchmark LGTM. But what's the behavior when running `cargo test --package tests -- --test bench_txn_store_rocksdb_inc`? I think that's the current behavior of CI.

[the remark this answers is not in the document]
Comment (tabokie) [reply components/test_util/src/runner.rs:53]: It still works. It seems benches are exactly the same as tests unless `--bench` is provided.

File: components/test_util/src/runner.rs [hunk 2]
```diff
@@ -31,41 +31,40 @@ impl TestHook for Nope {
 }
 
 struct CaseLifeWatcher<H: TestHook> {
-    name: String,
     hook: H,
 }
 
 impl<H: TestHook + Send + 'static> CaseLifeWatcher<H> {
-    fn new(name: String, mut hook: H) -> CaseLifeWatcher<H> {
-        debug!("case start"; "name" => &name);
```
Comment (BusyJay) [inline components/test_util/src/runner.rs:40]: Why delete it?

Comment (tabokie) [reply components/test_util/src/runner.rs:40]: Don't even remember. Reverted.

**Review by BusyJay — APPROVED** [review]

File: components/test_util/src/runner.rs [hunk 3]
```diff
@@ -57,15 +57,15 @@ pub fn run_test_with_hook(cases: &[&TestDescAndFn], hook: impl TestHook + Send +
         .iter()
         .map(|case| {
             let name = case.desc.name.as_slice().to_owned();
-            let h = hook.clone();
+            let hook = hook.clone();
```
Comment (tabokie) [inline components/test_util/src/runner.rs:60]:
```suggestion
            let hook = hook.clone();
```

**Pull request merged**